docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,474 @@
+
+In the next section, we'll show how to generalize the model to handle the absense of type annotations.
+
+Continued example: Dealing with untyped code
+--------------------------------------------
+
+Suppose we want the model from above to detect the sink in this snippet:
+
+.. code-block:: js
+
+  import { getConnection } from "@example/db";
+  let connection = getConnection();
+  connection.query(q); // <-- add 'q' as a SQL injection sink
+
+There is no type annotation on **connection**, and there is no indication of what **getConnection()** returns.
+Using a **typeModel** tuple we can tell our model that this function returns an instance of **mysql.Connection**:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: typeModel
+      data:
+        - ["mysql.Connection", "@example/db", "Member[getConnection].ReturnValue"]
+
+To break this down:
+
+- Since we're providing type information, we add a tuple to the **typeModel** extension point.
+- The first column, **"mysql.Connection"**, names the type that we're adding a new definition for.
+- The second column, **"@example/db"**, begins the search at imports of the hypothetical NPM package **@example/db**.
+- **Member[getConnection]** selects references to the **getConnection** member from that package.
+- **ReturnValue** selects the return value from a call to that member.
+
+The new model states that the return value of **getConnection()** has type **mysql.Connection**.
+Combining this with the sink model we added earlier, the sink in the example is detected by the model.
+
+The mechanism used here is how library models work for both TypeScript and plain JavaScript.
+A good library model contains **typeModel** tuples to ensure it works even in codebases without type annotations.
+For example, the **mysql** model that is included with the CodeQL JS analysis includes this type definition (among many others):
+
+.. code-block:: yaml
+
+  - ["mysql.Connection", "mysql", "Member[createConnection].ReturnValue"]
+
+Example: Adding flow through 'decodeURIComponent'
+-------------------------------------------------
+
+In this example, we'll show how to add flow through calls to `decodeURIComponent`:
+
+.. code-block:: js
+
+  let y = decodeURIComponent(x); // add taint flow from 'x' to 'y'
+
+This flow is already recognized by the CodeQL JS analysis, but this is how it could be added with an extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: summaryModel
+      data:
+        - [
+            "global",
+            "Member[decodeURIComponent]",
+            "Argument[0]",
+            "ReturnValue",
+            "taint",
+          ]
+
+To break this down:
+
+- Since we're adding flow *through* a function call, we add a tuple to the **summaryModel** extension point.
```

Comment:
```suggestion
- Since we're adding flow through a function call, we add a tuple to the **summaryModel** extension point.
```

Minor but italics can be confusing for some users